tokio/tests/sync_mpsc.rs

```diff
@@ -943,3 +943,233 @@ async fn test_tx_capacity() {
+                }
+                ActorMessage::SelfMessage { .. } => {
+                    self.received_self_msg = true;
+                }
+            }
+        }
+
+        async fn send_message_to_self(&mut self) {
+            let msg = ActorMessage::SelfMessage {};
+
+            let sender = self.sender.clone();
+
+            // cannot move self.sender here
+            if let Some(sender) = sender.upgrade() {
+                let _ = sender.send(msg);
+                self.sender = sender.downgrade();
+            }
+        }
+
+        async fn run(&mut self) {
+            let mut i = 0;
+            while let Some(msg) = self.receiver.recv().await {
+                self.handle_message(msg);
+
+                if i == 0 {
+                    self.send_message_to_self().await;
+                }
+
+                i += 1
+            }
+
+            assert!(self.received_self_msg);
+        }
+    }
+
+    #[derive(Clone)]
+    pub struct MyActorHandle {
+        sender: mpsc::UnboundedSender<ActorMessage>,
+    }
+
+    impl MyActorHandle {
+        pub fn new() -> (Self, MyActor) {
+            let (sender, receiver) = mpsc::unbounded_channel();
+            let actor = MyActor::new(receiver, sender.clone().downgrade());
+
+            (Self { sender }, actor)
+        }
+
+        pub async fn get_unique_id(&self) -> u32 {
+            let (send, recv) = oneshot::channel();
+            let msg = ActorMessage::GetUniqueId { respond_to: send };
+
+            // Ignore send errors. If this send fails, so does the
+            // recv.await below. There's no reason to check the
+            // failure twice.
+            let _ = self.sender.send(msg);
+            recv.await.expect("Actor task has been killed")
+        }
+    }
+
+    let (handle, mut actor) = MyActorHandle::new();
+
+    let actor_handle = tokio::spawn(async move { actor.run().await });
+
+    let _ = tokio::spawn(async move {
+        let _ = handle.get_unique_id().await;
+        drop(handle);
+    })
+    .await;
+
+    let _ = actor_handle.await;
+}
+
+// Tests that no pending messages are put onto the channel after `Rx` was
+// dropped.
+//
+// Note: After the introduction of `UnboundedWeakSender`, which internally
+// used `Arc` and doesn't call a drop of the channel after the last strong
+// `UnboundedSender` was dropped while more than one `UnboundedWeakSender`
+// remains, we want to ensure that no messages are kept in the channel, which
+// were sent after the receiver was dropped.
+#[tokio::test]
+async fn test_msgs_dropped_on_unbounded_rx_drop() {
+    let (tx, mut rx) = mpsc::unbounded_channel();
+
+    tx.send(Msg {}).unwrap();
+    tx.send(Msg {}).unwrap();
+
+    // This msg will be pending and should be dropped when `rx` is dropped
+    let sent = tx.send(Msg {});
+
+    let _ = rx.recv().await.unwrap();
+    let _ = rx.recv().await.unwrap();
+
+    sent.unwrap();
+
+    drop(rx);
+
+    assert_eq!(NUM_DROPPED.load(Acquire), 3);
```

Comment:
Is this shared by tests? That's not so good because they run in parallel and access the same globals.

Reply:
I just made separate version for unbounded tests